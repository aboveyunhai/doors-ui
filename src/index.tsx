export * from './components/door';
export * from './utils/index';
