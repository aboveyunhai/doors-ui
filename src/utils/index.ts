export * from './isLazy';
export * from './use-callback-ref';
export * from './use-controllable';
export * from './use-disclosure';
export * from './use-id';
export * from './use-safe-layout-effect';
export * from './use-outside-click';

export function isNumber(value: any): value is number {
  return typeof value === 'number';
}

export function isFunction<T extends Function = Function>(
  value: any
): value is T {
  return typeof value === 'function';
}

export function runIfFn<T, U>(
  valueOrFn: T | ((...fnArgs: U[]) => T),
  ...args: U[]
): T {
  return isFunction(valueOrFn) ? valueOrFn(...args) : valueOrFn;
}

export type FunctionArguments<T extends Function> = T extends (
  ...args: infer R
) => any
  ? R
  : never;

export function callAllHandlers<T extends (event: any) => void>(
  ...fns: (T | undefined)[]
) {
  return function func(event: FunctionArguments<T>[0]) {
    fns.some((fn) => {
      fn?.(event);
      return event?.defaultPrevented;
    });
  };
}
