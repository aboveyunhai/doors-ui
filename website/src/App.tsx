import React from 'react';
import 'doors-ui/src/globals.css'
import { Door, useDisclosure } from 'doors-ui';
import { LoremIpsum } from 'lorem-ipsum';

function App() {
  const disclosurePropsOn = useDisclosure({ defaultIsOpen: true });
  const disclosurePropsOff = useDisclosure({ defaultIsOpen: false });
  const lorem = new LoremIpsum({
    sentencesPerParagraph: {
      max: 8,
      min: 4,
    },
    wordsPerSentence: {
      max: 16,
      min: 4,
    },
  });

  return (
    <div
      style={{
        width: '100%',
        height: '100%',
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
      }}
    >
      <button onClick={disclosurePropsOn.onToggle}>default on</button>
      <Door
        title={'A'}
        disclosureProps={disclosurePropsOn}
        lessThanScreen={true}
      >
        default is open
      </Door>
      <button onClick={disclosurePropsOff.onToggle}>default off</button>
      <Door
        title={'B'}
        disclosureProps={disclosurePropsOff}
        lessThanScreen={true}
      >
        {lorem.generateParagraphs(7)}
      </Door>
    </div>
  );
}

export default App;
