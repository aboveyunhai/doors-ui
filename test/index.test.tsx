import * as React from 'react';
import * as ReactDOM from 'react-dom';
import * as renderer from 'react-test-renderer';
import { Door, useDisclosure } from '../src/index';

describe('door', () => {
  const DoorTest = () => {
    const disclosureProps = useDisclosure();
    return (
      <Door title={'test'} disclosureProps={disclosureProps}>
        test
      </Door>
    );
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<DoorTest />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('is close', () => {
    const component = renderer.create(<DoorTest />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
