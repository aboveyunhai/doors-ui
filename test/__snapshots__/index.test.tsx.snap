// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`door is close 1`] = `
<div
  data-projection-id={1}
  onPointerDownCapture={[Function]}
  style={
    Object {
      "backgroundColor": "white",
      "border": "1px solid rgba(186, 188, 190, 0.5)",
      "height": "400px",
      "left": 0,
      "minWidth": "200px",
      "opacity": 0,
      "position": "fixed",
      "top": 0,
      "transform": "none",
      "visibility": "hidden",
      "width": "300px",
      "zIndex": 19999,
    }
  }
/>
`;
